remove dependency on fcrepo-mint

acoburn opened the pull request:
See: https://jira.duraspace.org/browse/FCREPO-1642

escowles commented:
This replicates the current behavior, but there's also an outstanding ticket for making the event ids configurable: https://jira.duraspace.org/browse/FCREPO-1568

Can we inject a Supplier<String>, allowing us to wire up whatever kind of ID minter was desired?  This would also remove the dependency on fcrepo-mint.

ajs6f commented:
@escowles How do you want to use a `Supplier`? To supply a minter or to supply ids? Keep in mind that the basic `Supplier` type doesn't guarantee that you get the same thing each time, so supplying ids makes sense, but not supplying minters.

escowles commented:
@ajs6f I think it's stripping out the angle brackets: I meant to say Supplier&lt;String&gt; -- so a supplier of ids.  I think it makes sense to use the same approach we're using for the main repository ids in https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-http-commons/src/main/java/org/fcrepo/http/commons/AbstractResource.java#L101

ajs6f commented:
Holy crap! That's no good. `Supplier` doesn't guarantee that it _won't_ give you the same thing over and over again. I hope I didn't do that in `AbstractResource`, but I think I might have absent-mindedly done that. If we want to use `Supplier` to produce a stream of unique things, we have to subtype it.

escowles commented:
So maybe we should add a PidMinter interface to fcrepo-kernel-api, and then inject fcrepo-mint instances in the fcrepo-http-commons repository config?

ajs6f commented:
I don't think we should be as specific as `PidMinter` because we don't need to be. `UniqueValueSupplier` would do it. Then we could toast fcrepo-mint, right? We could offer a sidecar extension with the HTTP minter in it.

escowles commented:
`UniqueValueSupplier<String>` sounds good to me.  Do you mean remove fcrepo-mint altogether?  Do the UUID impls. get moved to `fcrepo-kernel-modeshape` or somewhere else?

ajs6f commented:
Hey, man, this is java 8! The UUID impl becomes default method bodies on the interface itself.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/observer/FedoraEvent.java:
```diff
@@ -64,7 +61,7 @@
     public FedoraEvent(final Event e) {
         checkArgument(e != null, "null cannot support a FedoraEvent!");
-        eventID = pidMinter.get();
+        eventID = randomUUID().toString();
```
Instead of hard-coding `randomUUID()` here, is there a reason why an `UniqueValueSupplier<String>`, as suggested in the code review comments, was not injected?

As a note, `randomUUID().toString()` is not equivalent to the previous `UUIDPathMinter.get()`. `UUIDPathMinter` provides a pairtree path with a UUID leaf element, whereas `randomUUID` simply provides a UUID.
The reason a full pairtree path is important is because the `InternalAuditor` uses the `FedoraEvent#eventID` as the identifier name at which an audit record resource is created:
https://github.com/fcrepo4-labs/fcrepo-audit/blob/master/src/main/java/org/fcrepo/audit/InternalAuditor.java#L211

The pairtree path provided by the `UUIDPathMinter` gets around the performance limitation of too many children of a given resource.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/observer/FedoraEvent.java:
```diff
@@ -64,7 +61,7 @@
      */
     public FedoraEvent(final Event e) {
         checkArgument(e != null, "null cannot support a FedoraEvent!");
```
This should be Java 8 `Objects::requireNonNull`.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/observer/FedoraEvent.java:
```diff
@@ -27,14 +28,12 @@
-import org.fcrepo.mint.UUIDPathMinter;
 
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
```
These Guava types should be replaced with java 8 types `Function` and `StringJoiner` (or potentially `String::join` or `Collectors::joining`).

acoburn replied:
I completely agree. This PR, however, was focused entirely on removing the dependency on fcrepo-mint, so doing all the java8 refactoring seemed out of scope.

awoods replied:
Agreed, @acoburn... but thanks for the tips, @ajs6f!

acoburn commented:
Updated PR with an injectable `UniqueValueSupplier` that subclasses `Supplier<String>`. Now neither of the `fcrepo-kernel-*` modules depend on `fcrepo-mint`.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/functions/UniqueValueSupplier.java:
```diff
@@ -0,0 +1,49 @@
+ * @author awoods
+ * @author acoburn
+ */
+public interface UniqueValueSupplier extends Supplier<String> {
```
This is way more specific than the name `UniqueValueSupplier` would imply. How about `HierarchicalIdentifierSupplier` < `UniqueValueSupplier`, with `UniqueValueSupplier` simply guaranteeing the uniqueness of results (no default impl) and `HierarchicalIdentifierSupplier` bringing the "minter that creates hierarchical IDs from a UUID" semantic with the default impl?

acoburn replied:
That works for me. I can add the additional interface.

ajs6f replied:
Right on.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/functions/UniqueValueSupplier.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    default public String get() {
+        final int defaultLength = 2;
+        final int defaultCount = 4;
```
Is there some reason these are local variables?

acoburn replied:
can you have private fields on an interface?

ajs6f replied:
Why `private`? These are defaults the effects of which are entirely obvious-- why not `(public) static`?

acoburn replied:
That makes sense -- I had been trying not to leak implementation details, but in this case, that's probably a good thing to advertise.

ajs6f replied:
Let it all hang out. There is no place for shame or modesty in Fedora.

acoburn commented:
added an interface for `HierarchicalIdentifierSupplier` < `UniqueValueSupplier` and moved the default impl to `HierarchicalIdentifierSupplier`, as @ajs6f suggested

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/functions/UniqueValueSupplier.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.function.Supplier;
+
+/**
+ * Unique value minter that creates unique IDs
```
This is now the wrong comment. This type just guarantees the uniqueness of its supplied values. It has nothing to do with minting IDs.

acoburn replied:
The comment has been updated.

awoods commented:
This looks good. @escowles, you will note that the issue of injecting a `pidMinter` implementation into `FedoraEvent`s is still outstanding, for the same reasons mentioned in the ticket:
https://jira.duraspace.org/browse/FCREPO-1568

If @escowles or @ajs6f are comfortable with this current PR as well, I will go ahead and squash/merge.

escowles commented:
:+1:  -- looks like a good improvement, and we can figure out the injection separately.

acoburn commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/observer/FedoraEvent.java:
```diff
@@ -54,8 +54,7 @@
     private Set<String> eventProperties = new HashSet<>();
-
-    private static final Supplier<String> pidMinter = new UUIDPathMinter();
+    private UniqueValueSupplier pidMinter = new DefaultPathMinter();
```
shouldn't this be `static final`?

awoods replied:
yes

awoods replied:
If you push one more commit with the addition of `static` we can close this PR.

acoburn replied:
Commit pushed (actually two). The nested classes had to be made static in order to use them on a static field.

acoburn commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/rdf/JcrRdfTools.java:
```diff
@@ -108,7 +108,7 @@
     private static final Model m = createDefaultModel();
 
-    private static final Supplier<String> pidMinter =  new UUIDPathMinter();
+    private final UniqueValueSupplier pidMinter =  new DefaultPathMinter();
```
And shouldn't this be `static`?

awoods replied:
yes

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/functions/UniqueValueSupplier.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author acoburn
+ */
+public interface UniqueValueSupplier extends Supplier<String> {
```
Since this PR keeps staying open, I keep looking at it...
Why is this new class `UniqueValueSupplier` in the `org.fcrepo.kernel.api.services.functions` package?

acoburn replied:
I put them in `o.f.k.a.services.functions` because they are different types of `Supplier` interfaces, which, itself provides a functional interface. That's the only reason.... If you'd prefer they go in `o.f.k.a.identifiers`, that's fine by me.

awoods replied:
Tying it back to `Supplier` is rationale enough for me.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/functions/HierarchicalIdentifierSupplier.java:
```diff
@@ -0,0 +1,49 @@
+ * @author awoods
+ * @author acoburn
+ */
+public interface HierarchicalIdentifierSupplier extends UniqueValueSupplier {
```
Why is this new class `HierarchicalIdentifierSupplier` in the `org.fcrepo.kernel.api.services.functions` package?

awoods commented:
Resolved with: https://github.com/fcrepo4/fcrepo4/commit/f7daf307912d78cc57098a74c2b04c2daf5d54e6
